# Import Libraries

In [ ]:
# First we'll import the basic libraries that we'll use in this project
import pandas as pd
import numpy as np
from numpy import cov
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.patches import Ellipse
import matplotlib.transforms as transforms

import ipywidgets as widgets
from ipywidgets import interact

# Dataset Description

In [ ]:
# Load the dataset we'll use here
data = pd.read_csv('/content/insurance.csv')
data.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


Based on `info()` we can conclude that our dataset consist of 1338 rows and 6 features. So far it seems that there is no missing value there, but we will still make sure later.

# Data Cleaning

## Handling Duplicated Values

In [ ]:
data[data.duplicated(keep=False)]

age   sex    bmi  children smoker     region    charges
195   19  male  30.59         0     no  northwest  1639.5631
581   19  male  30.59         0     no  northwest  1639.5631

In [ ]:
data = data.drop_duplicates(keep="first")
data[data.duplicated(keep=False)]

Empty DataFrame
Columns: [age, sex, bmi, children, smoker, region, charges]
Index: []

**We have handled the duplicated value!**

## Handling the Missing Value

In [ ]:
data.isna().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

**Wow there is no missing value!**

## Handling the Inconsistent format

In [ ]:
data["age"].unique()

array([19, 18, 28, 33, 32, 31, 46, 37, 60, 25, 62, 23, 56, 27, 52, 30, 34,
       59, 63, 55, 22, 26, 35, 24, 41, 38, 36, 21, 48, 40, 58, 53, 43, 64,
       20, 61, 44, 57, 29, 45, 54, 49, 47, 51, 42, 50, 39])

In [ ]:
data["sex"].unique()

array(['female', 'male'], dtype=object)

In [ ]:
data["bmi"].unique()

array([27.9  , 33.77 , 33.   , 22.705, 28.88 , 25.74 , 33.44 , 27.74 ,
       29.83 , 25.84 , 26.22 , 26.29 , 34.4  , 39.82 , 42.13 , 24.6  ,
       30.78 , 23.845, 40.3  , 35.3  , 36.005, 32.4  , 34.1  , 31.92 ,
       28.025, 27.72 , 23.085, 32.775, 17.385, 36.3  , 35.6  , 26.315,
       28.6  , 28.31 , 36.4  , 20.425, 32.965, 20.8  , 36.67 , 39.9  ,
       26.6  , 36.63 , 21.78 , 30.8  , 37.05 , 37.3  , 38.665, 34.77 ,
       24.53 , 35.2  , 35.625, 33.63 , 28.   , 34.43 , 28.69 , 36.955,
       31.825, 31.68 , 22.88 , 37.335, 27.36 , 33.66 , 24.7  , 25.935,
       22.42 , 28.9  , 39.1  , 36.19 , 23.98 , 24.75 , 28.5  , 28.1  ,
       32.01 , 27.4  , 34.01 , 29.59 , 35.53 , 39.805, 26.885, 38.285,
       37.62 , 41.23 , 34.8  , 22.895, 31.16 , 27.2  , 26.98 , 39.49 ,
       24.795, 31.3  , 38.28 , 19.95 , 19.3  , 31.6  , 25.46 , 30.115,
       29.92 , 27.5  , 28.4  , 30.875, 27.94 , 35.09 , 29.7  , 35.72 ,
       32.205, 28.595, 49.06 , 27.17 , 23.37 , 37.1  , 23.75 , 28.975,
      

In [ ]:
data["children"].unique()

array([0, 1, 3, 2, 5, 4])

In [ ]:
data["smoker"].unique()

array(['yes', 'no'], dtype=object)

In [ ]:
data["region"].unique()

array(['southwest', 'southeast', 'northwest', 'northeast'], dtype=object)

In [ ]:
data["charges"].unique()

array([16884.924 ,  1725.5523,  4449.462 , ...,  1629.8335,  2007.945 ,
       29141.3603])

In [ ]:
data.head(10)

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520
5   31  female  25.740         0     no  southeast   3756.62160
6   46  female  33.440         1     no  southeast   8240.58960
7   37  female  27.740         3     no  northwest   7281.50560
8   37    male  29.830         2     no  northeast   6406.41070
9   60  female  25.840         0     no  northwest  28923.13692

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1337 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1337 non-null   int64  
 1   sex       1337 non-null   object 
 2   bmi       1337 non-null   float64
 3   children  1337 non-null   int64  
 4   smoker    1337 non-null   object 
 5   region    1337 non-null   object 
 6   charges   1337 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 83.6+ KB


# Descriptive Analytics

In [ ]:
data.describe(include="all")

age   sex          bmi     children smoker     region  \
count   1337.000000  1337  1337.000000  1337.000000   1337       1337   
unique          NaN     2          NaN          NaN      2          4   
top             NaN  male          NaN          NaN     no  southeast   
freq            NaN   675          NaN          NaN   1063        364   
mean      39.222139   NaN    30.663452     1.095737    NaN        NaN   
std       14.044333   NaN     6.100468     1.205571    NaN        NaN   
min       18.000000   NaN    15.960000     0.000000    NaN        NaN   
25%       27.000000   NaN    26.290000     0.000000    NaN        NaN   
50%       39.000000   NaN    30.400000     1.000000    NaN        NaN   
75%       51.000000   NaN    34.700000     2.000000    NaN        NaN   
max       64.000000   NaN    53.130000     5.000000    NaN        NaN   

             charges  
count    1337.000000  
unique           NaN  
top              NaN  
freq             NaN  
mean    13279.121487  
std     12110.359656  
min      1121.873900  
25%      4746.344000  
50%      9386.161300  
75%     16657.717450  
max     63770.428010

Based on `describe()`, we can conclude that:
- Our oldest respondent is 64 years old, with average 39 years old
- More than 50% of our respondent are male
- Our respondent highest `bmi` is 53
- Most of our respondents have only 1 child, but few/some of them have 5 `children`
- 80% of our respondents are a non-`smoker`
- 30% of our respondents live in Southeast `region` and
- Our respondent average `charges` on their insurance is 13279.121487.

# Analysis

In [ ]:
data[["smoker", "bmi"]].groupby(["smoker"]).agg(["max", "min", "mean"]).reset_index()

smoker    bmi                   
            max     min       mean
0     no  53.13  15.960  30.651853
1    yes  52.58  17.195  30.708449

In [ ]:
data[["smoker", "age", "sex"]].groupby(["smoker", "sex"]).agg(["max", "min", "mean", "sum"]).reset_index()

smoker     sex age                      
                 max min       mean    sum
0     no  female  64  18  39.691042  21711
1     no    male  64  18  39.100775  20176
2    yes  female  64  18  38.608696   4440
3    yes    male  64  18  38.446541   6113

In [ ]:
data[["smoker", "charges"]].groupby(["smoker"]).agg(["max", "min", "mean", "sum"]).reset_index()

smoker      charges                                        
                  max         min          mean           sum
0     no  36910.60803   1121.8739   8440.660307  8.972422e+06
1    yes  63770.42801  12829.4551  32050.231832  8.781764e+06

## Discrete Variable

In [ ]:
data[["sex", "charges"]].groupby(["sex"]).agg(["max", "min", "mean"]).reset_index()

sex      charges                                       
                   max        min          mean           sum
0  female  63770.42801  1607.5101  12569.578844  8.321061e+06
1    male  62592.87309  1121.8739  13974.998864  9.433124e+06

In [ ]:
data[["region", "charges"]].groupby(["region"]).agg(["max", "min", "mean", "sum"]).reset_index()

region      charges                                       
                      max        min          mean           sum
0  northeast  58571.07448  1694.7964  13406.384516  4.343669e+06
1  northwest  60021.39897  1621.3402  12450.840844  4.034072e+06
2  southeast  63770.42801  1121.8739  14735.411438  5.363690e+06
3  southwest  52590.82939  1241.5650  12346.937377  4.012755e+06

Dengan kovarians tidak mudah ditafsirkan secara kualitatif, sehingga kita perlu menghitung korelasi.

## Correlation

In [ ]:
data.corr()

<ipython-input-25-c44ded798807>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.corr()


age       bmi  children   charges
age       1.000000  0.109344  0.041536  0.298308
bmi       0.109344  1.000000  0.012755  0.198401
children  0.041536  0.012755  1.000000  0.067389
charges   0.298308  0.198401  0.067389  1.000000